### ISSUE TITLE
Unsound result with recursion

### ISSUE THREAD

**Issue by jtoman** [jayhorn/jayhorn#154]
Hello,

I'm a researcher looking at JayHorn. Experimenting with some benchmarks, I found that the following program verifies when it should not. The benchmark recurses through rec twice, calling neg twice, and then increments the contents of Ptr before returning. The assertion definitely fails (as confirmed by running the program with a concrete value). I am running this with Java 1.8.0_212-8u212-b03-0ubuntu1.16.04.1-b03 and Jayhorn built from revision c09739e13540e99fa8bd195e4e5e93814b22252c.

```java
class FalseNegative {
	public static class Ptr {
		public int a;
	}

	public static void neg(Ptr p) {
		p.a = -p.a;
	}

	public static void rec(Ptr p, boolean b) {
		if(b) {
			neg(p);
			p.a++;
		} else {
			neg(p);
			rec(p, true);
		}
	}
	
	public static void main(String[] args) {
		Ptr p = new Ptr();
		p.a = Havoc_Class.havoc_int();
		int d = p.a;
		rec(p, false);
		assert d == p.a;
	}
}
```

**Comment by lememta** [comment]
Can you post how are you running JayHorn? Which options are you using>

**Comment by jtoman** [comment]
`java -jar jayhorn/build/libs/jayhorn.jar -j ~/path/to/classes`

The above folder does not contain an class file for Havoc_Class, which I believe is sufficient to trigger  stubbing behavior.

**Comment by pruemmer** [comment]
Thanks for the bug report! I can reproduce this on my machine; looking at the CFG (option -cfg), it seems that the assertion is simplified away completely. The bug might be due to some of the recent changes in the constant propagator and partial evaluator; I’m looking into it.

-- 
Philipp Rümmer, PhD, Docent (Associate Professor)
Department of Information Technology, Uppsala University

**Comment by pruemmer** [comment]
This should be fixed by the latest commit, 4f998aa
